mpc-spec/src/party.rs

```diff
@@ -0,0 +1,164 @@
+/// - `parties`: All other parties' sender handles, ordered by their `id`s
+/// - `id`: The owning parties `id`
+#[allow(dead_code)] // TODO: Remove this later.
+pub struct ChannelConfig {
+    pub(crate) listen: Receiver<MPCMessage>,
+    pub(crate) evaluator: Sender<MPCMessage>,
+    pub(crate) subprotocol: Sender<MPCMessage>,
+    pub(crate) parties: Vec<Sender<MPCMessage>>,
+    pub(crate) id: usize,
+}
+
+/// A struct defining protocol party state during a protocol execution.
+#[allow(dead_code)] // TODO: Remove this later.
+pub struct Party {
+    /// The parties numeric identifier
+    id: usize,
+    /// The channel configuration for communicating to other protocol parties
+    channels: ChannelConfig,
+    /// The global MAC key for authenticating bit shares
+    global_mac_key: Option<MacKey>,
+    /// The circuit to be evaluated during the multi-party computation
+    circuit: Circuit,
+    /// A source of random bytes and bits local to the party
+    entropy: Randomness,
+    /// Tracks the current phase of protocol execution
+    current_phase: ProtocolPhase,
+}
+
+#[allow(dead_code)] // TODO: Remove this later.
+impl Party {
+    /// Initialize an MPC party.
+    pub fn new(channels: ChannelConfig, circuit: &Circuit, entropy: Randomness) -> Self {
+        Self {
+            id: channels.id,
+            channels,
+            global_mac_key: None,
+            circuit: circuit.clone(),
+            entropy,
+            current_phase: ProtocolPhase::PreInit,
+        }
+    }
+
+    /// Return `true`, if the party is the designated circuit evaluator.
+    fn is_evaluator(&self) -> bool {
+        self.id == 0
+    }
+
+    /// Send a message to the subprotocol channel.
+    fn subprotocol(&mut self, message: MPCMessage) {
+        self.channels.subprotocol.send(message).unwrap();
+        let msg = self.channels.listen.recv().unwrap();
+        self.process_message(msg);
+    }
+
+    /// Send a message to the evaluator.
+    fn send_to_evaluator(&mut self, message: MPCMessage) {
+        self.channels.evaluator.send(message).unwrap();
+    }
+
+    /// Participate in a full communication round between all parties.
+    ///
+    /// Communication turn-order is determined by the parties `id`, i.e. earlier
+    /// `id`s send before later `id`s.
+    fn full_round(&mut self, round_function: fn(usize, usize) -> MPCMessage) {
+        let num_parties = self.channels.parties.len();
+
+        // Expect earlier parties' messages.
+        for _i in 0..self.id {
+            let msg = self.channels.listen.recv().unwrap();
+            self.process_message(msg);
+        }
+
+        // All earlier messages have been received, so it is the parties' turn
+        // to send messages to everyone, except itself.
+        for i in 0..num_parties {
+            if i == self.id {
+                continue;
+            }
+
+            self.channels.parties[i % num_parties]
+                .send(round_function(self.id, i % num_parties))
```

Comment:
Why the modular reductions?

Reply:
You're right, they are not necessary at all since `i in 0.. num_parties`. :+1: